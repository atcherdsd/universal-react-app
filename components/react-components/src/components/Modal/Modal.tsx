import ExitModal from 'components/ExitModal/ExitModal';
import { Data } from 'components/SearchBar/SearchBar';
import React from 'react';
import './Modal.css';

type HandleResult = { (): void };

interface IModalProps {
  data: Data;
  handleResult: HandleResult;
}

function Modal(props: IModalProps): JSX.Element {
  const date = props.data.published_at.slice(0, 10);

  const decodeHtmlCharCodes = (str: string) =>
    str
      .replace(/(&#(\d+);)/g, (_match, _capture, charCode: string) =>
        String.fromCharCode(+charCode)
      )
      .replace(/&rsquo;/g, '’')
      .replace(/&nbsp;/g, ' ')
      .replace(/&hellip;/g, '…')
      .replace(/&mdash;/g, '—')
      .replace(/&amp;/g, '&')
      .replace(/&reg;/g, '®')
      .replace(/&trade;/g, '™');

  return (
    <div className="Modal-container">
      <div className="Modal-content__container" onClick={(event) => event.stopPropagation()}>
        <div className="Modal-main-content">
          <div className="Modal-content__header">
            <div className="Modal-source">Source: {decodeHtmlCharCodes(props.data.source)}</div>
            <div className="Modal-category">News category: {props.data.category}</div>
            <div className="Modal-country">Country: {props.data.country}</div>
            <div className="Modal-language">{props.data.language}</div>
          </div>
          <div className="Modal-content__main">
            {props.data.image && (
              <div className="Modal-image__container">
                <img className="Modal-image" src={props.data.image} alt="News image" />
              </div>
            )}
            <div
              className="Modal-content__info"
              style={props.data.image ? { width: '60%' } : { width: '100%' }}
            >
              <div className="Modal-title">{decodeHtmlCharCodes(props.data.title)}</div>
              <div className="Modal-description">{decodeHtmlCharCodes(props.data.description)}</div>
            </div>
          </div>
          <div className="Modal-content__footer">
            <div className="Modal-date">Publication date: {date}</div>
            {props.data.author && <div className="Modal-author">Author: {props.data.author}</div>}
            <a href={props.data.url} className="Modal-url" target="blank" rel="noreferrer">
              See in source
            </a>
          </div>
        </div>
        <div className="Modal-exit" onClick={props.handleResult}>
          <ExitModal />
        </div>
      </div>
    </div>
  );
}

export default Modal;
