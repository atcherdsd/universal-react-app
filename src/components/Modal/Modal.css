.Modal-container {
    border-radius: 1rem;
    background: wheat;
    width: 80%;
    max-width: 1200px;
    overflow: auto;
    max-height: 95vh;
}
.Modal-content__container {
    display: flex;
    justify-content: space-around;
    position: relative;
}
.Modal-exit-ico {
    padding: 0.5rem;
    background: rgb(238, 204, 159);
    position: absolute;
    right: 0;
    top: 0;
    border-radius: 0 0 0 1rem;
}
.Modal-exit:hover {
    background: tan;
}

.Modal-image {
    max-height: 11rem;
    transform: translate(-50%, -50%);
    position: relative;
    top: 50%;
    left: 50%;
}

.Modal-main-content {
    cursor: auto;
    padding: 2.5rem;
}

.Modal-title,
.Modal-description {
    padding: 1rem;
}
.Modal-title {
    font-size: 1.5rem;
}
.Modal-content__header,
.Modal-content__main,
.Modal-content__footer {
    display: flex;
    flex-wrap: wrap;
    padding: 1rem;
    gap: 1.5rem;
}
.Modal-content__header,
.Modal-content__footer {
    background-color: #f5f4f4;
    justify-content: space-between; 
}
.Modal-content__main {
    justify-content: center;
}

.Modal-content__info {
    display: flex;
    flex-direction: column;
    justify-content: center;
    flex-grow: 1;
}
.info__w60 {
    width: 60%;
}
.info__w100 {
    width: 100%;
}

.Modal-language {    
    background-color: #90bfe5;
    color: blue;
}
.Modal-description a,
.Modal-url {
    text-decoration: none;
}

@media(max-width: 1000px) {
    .Modal-content__container {
        flex-direction: column;
    }
    .Modal-image {
        transform: none;
        left: auto;
        top: auto;
    }
    .Modal-content__header,
    .Modal-content__footer {
        justify-content: center;
    }
    .Modal-title,
    .Modal-description {
        padding: 1rem 0;
    }
    .Modal-content__main{
        padding: 1rem 0;
    }
}
