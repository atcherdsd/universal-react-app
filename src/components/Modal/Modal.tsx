import IcoModalExit from 'components/IcoModalExit/IcoModalExit';
import { IModalProps } from 'components/utilities/interfaces';
import { decodeHtmlCharCodes } from 'components/utilities/utils';
import React, { RefObject, useEffect, useRef } from 'react';
import './Modal.css';

function Modal(props: IModalProps): JSX.Element {
  const descriptionDiv = useRef() as RefObject<HTMLDivElement>;

  useEffect(() => {
    const replaceLink = (text: string): string => {
      descriptionDiv.current!.innerHTML = text;
      return decodeHtmlCharCodes(descriptionDiv.current!.innerHTML);
    };
    replaceLink(props.data.description);
  }, [props.data.description]);

  return (
    <div className="Modal-container">
      <div className="Modal-content__container" onClick={(event) => event.stopPropagation()}>
        <div className="Modal-main-content">
          <div className="Modal-content__header">
            <div className="Modal-source">
              Source: {decodeHtmlCharCodes(props.data.source.name)}
            </div>
          </div>
          <div className="Modal-content__main">
            {props.data.image && (
              <div className="Modal-image__container">
                <img className="Modal-image" src={props.data.image} alt="News image" />
              </div>
            )}
            <div
              className={
                props.data.image
                  ? 'Modal-content__info  info__w60'
                  : 'Modal-content__info info__w100'
              }
            >
              <div className="Modal-title">{decodeHtmlCharCodes(props.data.title)}</div>
              <div className="Modal-description" ref={descriptionDiv}></div>
            </div>
          </div>
          <div className="Modal-content__footer">
            <div className="Modal-date">Publication date: {props.date}</div>
            <a href={props.data.url} className="Modal-url" target="blank" rel="noreferrer">
              See in source
            </a>
          </div>
        </div>
        <div className="Modal-exit-ico" onClick={props.handleResult}>
          <IcoModalExit />
        </div>
      </div>
    </div>
  );
}

export default Modal;
