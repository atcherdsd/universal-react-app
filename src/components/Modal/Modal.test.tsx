import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import Modal from './Modal';
// import SearchBar from 'components/SearchBar/SearchBar';
// import SearchResult from 'components/SearchResult/SearchResult';

describe('Modal component', () => {
  test('should render Modal component', () => {
    const snapshot = render(
      <Modal
        data={{
          author: '',
          title: '',
          description: '',
          urlToImage: '',
          publishedAt: '',
          source: {
            id: '',
            name: '',
          },
          url: '',
        }}
        handleResult={() => void {}}
      />
    );
    expect(snapshot).toMatchSnapshot();
  });
  it('onClick works', () => {
    const { unmount } = render(
      <Modal
        data={{
          author: '',
          title: '',
          description: '',
          urlToImage: '',
          publishedAt: '',
          source: {
            id: '',
            name: '',
          },
          url: '',
        }}
        handleResult={() => void {}}
      />
    );
    const exit = document.querySelector('.Modal-exit') as HTMLDivElement;
    exit.onclick = jest.fn();
    const onClick = exit.onclick;

    expect(screen.queryByText('See in source')).toBeInTheDocument();
    fireEvent.click(exit);
    expect(onClick).toHaveBeenCalledTimes(1);
    unmount();
    expect(screen.queryByText('See in source')).not.toBeInTheDocument();
  });
  test('Modal component do not have data', () => {
    expect(screen.queryByText(/home/i)).toBeNull();
    expect(screen.queryByRole('paragraph')).toBeNull();
  });
});
