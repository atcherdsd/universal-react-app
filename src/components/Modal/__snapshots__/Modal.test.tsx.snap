// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal component should render Modal component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="Modal-container"
      >
        <div
          class="Modal-content__container"
        >
          <div
            class="Modal-main-content"
          >
            <div
              class="Modal-content__header"
            >
              <div
                class="Modal-source"
              >
                Source: 
              </div>
            </div>
            <div
              class="Modal-content__main"
            >
              <div
                class="Modal-content__info info__w100"
              >
                <div
                  class="Modal-title"
                />
                <div
                  class="Modal-description"
                />
              </div>
            </div>
            <div
              class="Modal-content__footer"
            >
              <div
                class="Modal-date"
              >
                Publication date: 
              </div>
              <a
                class="Modal-url"
                href=""
                rel="noreferrer"
                target="blank"
              >
                See in source
              </a>
            </div>
          </div>
          <div
            class="Modal-exit"
          >
            <svg
              viewBox="0 0 12 12"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M.974 0L0 .974 5.026 6 0 11.026.974 12 6 6.974  11.026 12l.974-.974L6.974 6 12 .974 11.026 0 6 5.026z"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="Modal-container"
    >
      <div
        class="Modal-content__container"
      >
        <div
          class="Modal-main-content"
        >
          <div
            class="Modal-content__header"
          >
            <div
              class="Modal-source"
            >
              Source: 
            </div>
          </div>
          <div
            class="Modal-content__main"
          >
            <div
              class="Modal-content__info info__w100"
            >
              <div
                class="Modal-title"
              />
              <div
                class="Modal-description"
              />
            </div>
          </div>
          <div
            class="Modal-content__footer"
          >
            <div
              class="Modal-date"
            >
              Publication date: 
            </div>
            <a
              class="Modal-url"
              href=""
              rel="noreferrer"
              target="blank"
            >
              See in source
            </a>
          </div>
        </div>
        <div
          class="Modal-exit"
        >
          <svg
            viewBox="0 0 12 12"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M.974 0L0 .974 5.026 6 0 11.026.974 12 6 6.974  11.026 12l.974-.974L6.974 6 12 .974 11.026 0 6 5.026z"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
